eras/conway/impl/src/Cardano/Ledger/Conway/Governance/Proposals.hs

```diff
@@ -1,132 +1,451 @@
+deriving instance Era era => Ord (PForest PHierarchy era)
+deriving instance Era era => Generic (PForest PHierarchy era)
+deriving instance Era era => NoThunks (PForest PHierarchy era)
+deriving instance Era era => NFData (PForest PHierarchy era)
+deriving instance Era era => Default (PForest PHierarchy era)
+
+data Proposals era = Proposals
+  { pOMap :: !(OMap.OMap (GovActionId (EraCrypto era)) (GovActionState era))
+  , pRoots :: !(PForest PRoot era)
+  , pHierarchy :: !(PForest PHierarchy era)
+  }
+  deriving (Eq, Generic, NoThunks, NFData, Default)
+
+instance Era era => Show (Proposals era) where
+  show = flip proposalsShowDebug True
+
+pOMapL :: Lens' (Proposals era) (OMap.OMap (GovActionId (EraCrypto era)) (GovActionState era))
+pOMapL = lens pOMap $ \x y -> x {pOMap = y}
+
+pRootsL :: Lens' (Proposals era) (PForest PRoot era)
+pRootsL = lens pRoots $ \x y -> x {pRoots = y}
+
+pHierarchyL :: Lens' (Proposals era) (PForest PHierarchy era)
+pHierarchyL = lens pHierarchy $ \x y -> x {pHierarchy = y}
+
+instance EraPParams era => ToJSON (Proposals era) where
+  toJSON = toJSON . pOMap
+  toEncoding = toEncoding . pOMap
+
+proposalsAddAction :: forall era. GovActionState era -> Proposals era -> Maybe (Proposals era)
+proposalsAddAction gas ps =
+  -- TODO
+  -- case getLenses $ gas ^. gasActionL of
+  --   Nothing -> Just $ ps & pOMapL %~ (OMap.||> gas) -- TredsuryWithdrawals and InfoAction do not have their own trees
+  --   Just (lenzR, lenzH, parent) -> update lenzR lenzH parent
+  case gas ^. gasActionL of
+    ParameterChange parent _ ->
+      update pfrPParamUpdateL pfhPParamUpdateL parent
+    HardForkInitiation parent _ ->
+      update pfrHardForkL pfhHardForkL parent
+    TreasuryWithdrawals _ -> Just $ ps & pOMapL %~ (OMap.||> gas)
+    NoConfidence parent ->
+      update pfrCommitteeL pfhCommitteeL parent
+    UpdateCommittee parent _ _ _ ->
+      update pfrCommitteeL pfhCommitteeL parent
+    NewConstitution parent _ ->
+      update pfrConstitutionL pfhConstitutionL parent
+    InfoAction -> Just $ ps & pOMapL %~ (OMap.||> gas)
+  where
+    update ::
+      Lens' (PForest PRoot era) (PRoot (GovPurposeId p era)) ->
+      Lens' (PForest PHierarchy era) (PHierarchy (GovPurposeId p era)) ->
+      StrictMaybe (GovPurposeId p era) ->
+      Maybe (Proposals era)
+    update lenzR lenzH prev =
+      let newId = GovPurposeId $ gas ^. gasIdL
+          newNode parent = PNode parent Set.empty
+       in case prev of
+            SNothing ->
+              -- We only accept actions with parent `SNothing` when
+              -- there are no proposals enacted yet, which means that
+              -- the root has to be `Nothing`
+              if isNothing $ ps ^. pRootsL . lenzR . prRootL
+                then
+                  Just $
+                    ps
+                      & pOMapL %~ (OMap.||> gas)
+                      & pRootsL . lenzR . prChildrenL %~ Set.insert newId
```

Comment:
What do you think about having a few variables with meaningful names for some of these expressions that are reused often?  
Like: 
`pRootsL . lenzR . prChildrenL`
`pHierarchyL . lenzH . pHierarchyNTL` 

Maybe it will make it easier to understand the essence of this algorithm?

Reply:
Here `lenzR` and `lenzH` are polymorphic, and also passed into the scope from the calling site. Do you mean having a `let` binding like the following would help? I'd be happy to do this if you confirm that I understood your suggestion correctly. :grin: 
```haskell
let childrenOfRootL = pRootsL . lenzR . prChildrenL
```